File: foxpro-summarizer/llm/ollama_client.py
"""
Ollama REST API client for sending FoxPro code to Llama 3.1 for summarization.
"""
import json
import logging
import requests

from config import (
    OLLAMA_BASE_URL,
    OLLAMA_MODEL,
    OLLAMA_TIMEOUT,
    OLLAMA_TEMPERATURE,
    OLLAMA_NUM_PREDICT,
)

logger = logging.getLogger(__name__)


class OllamaClient:
    """Client for Ollama's local REST API."""

    def __init__(self, base_url=None, model=None):
        self.base_url = base_url or OLLAMA_BASE_URL
        self.model = model or OLLAMA_MODEL

    def test_connection(self):
        """Test if Ollama is running and the model is available."""
        try:
            resp = requests.get(f"{self.base_url}/api/tags", timeout=10)
            if resp.status_code == 200:
                data = resp.json()
                model_names = [m.get("name", "") for m in data.get("models", [])]
                # Check if model name matches (with or without :latest tag)
                found = any(
                    self.model in name or name.startswith(self.model)
                    for name in model_names
                )
                if found:
                    logger.info("Connected to Ollama. Model '%s' available.", self.model)
                    return True
                else:
                    logger.error(
                        "Model '%s' not found. Available: %s",
                        self.model,
                        ", ".join(model_names),
                    )
                    return False
        except requests.ConnectionError:
            logger.error("Cannot connect to Ollama at %s", self.base_url)
        except Exception as e:
            logger.error("Ollama connection test failed: %s", e)
        return False

    def generate(self, prompt):
        """Send a prompt to Ollama and return the response text."""
        payload = {
            "model": self.model,
            "prompt": prompt,
            "stream": False,
            "options": {
                "temperature": OLLAMA_TEMPERATURE,
                "num_predict": OLLAMA_NUM_PREDICT,
                "top_p": 0.9,
            },
        }

        resp = requests.post(
            f"{self.base_url}/api/generate",
            json=payload,
            timeout=OLLAMA_TIMEOUT,
        )
        resp.raise_for_status()
        data = resp.json()

        if "response" in data:
            return data["response"]
        raise ValueError(f"Unexpected response format: {json.dumps(data)[:500]}")

    def summarize_foxpro(self, code, context, include_python=None):
        """
        Summarize a FoxPro code block with rich context.
        Optionally includes the Python-converted equivalent for comparison.
        """
        prompt = self._build_prompt(code, context, include_python)
        return self.generate(prompt)

    def _build_prompt(self, code, context, python_code=None):
        """Build a detailed prompt for FoxPro analysis."""
        sections = [
            "You are a software analyst expert in Visual FoxPro (VFP) applications.",
            "Analyze the following FoxPro source code and provide a structured, "
            "human-readable summary.",
            "",
            f"Context: {context}",
            "",
            "Provide the following sections:",
            "",
            "## Use Case / Purpose",
            "What business function does this code serve? Describe in plain English.",
            "",
            "## Key Operations",
            "List the main operations (CRUD, calculations, validations, navigation).",
            "",
            "## Data Tables & Fields",
            "Which .dbf tables are opened/queried and what fields are used?",
            "",
            "## Business Rules & Validations",
            "What rules, constraints, or validations are enforced?",
            "",
            "## User Interface & Interactions",
            "What inputs, outputs, messages, or forms are involved?",
            "",
            "## Dependencies",
            "What other modules, procedures, forms, or class libraries does it call?",
            "",
            "## Access Control",
            "Are there any access level checks or permission-based logic?",
            "",
            "---",
            "",
            "FoxPro Source Code:",
            "```foxpro",
            code,
            "```",
        ]

        if python_code:
            sections.extend([
                "",
                "For reference, here is the Python equivalent (auto-converted):",
                "```python",
                python_code[:3000],
                "```",
            ])

        sections.extend([
            "",
            "Provide a clear, well-organized summary. Use bullet points and "
            "short paragraphs. Focus on business logic, not syntax.",
        ])

        return "\n".join(sections)

File: foxpro-analyzer/src/main/java/com/example/foxpro/llm/OllamaClient.java
package com.example.foxpro.llm;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

/**
 * Client for communicating with the Ollama REST API running locally.
 * Sends FoxPro code to Llama 3.1 for summarization.
 */
public class OllamaClient {

    private static final Logger logger = LoggerFactory.getLogger(OllamaClient.class);
    private static final MediaType JSON_MEDIA = MediaType.get("application/json");

    private final String baseUrl;
    private final String model;
    private final OkHttpClient httpClient;
    private final Gson gson;

    /**
     * @param baseUrl Ollama API base URL (default: http://localhost:11434)
     * @param model   Model name (e.g., "llama3.1")
     */
    public OllamaClient(String baseUrl, String model) {
        this.baseUrl = baseUrl;
        this.model = model;
        this.gson = new Gson();

        // Longer timeouts since LLM inference can take time, especially for large code blocks
        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(300, TimeUnit.SECONDS)  // 5 min for large responses
                .writeTimeout(30, TimeUnit.SECONDS)
                .build();
    }

    /**
     * Default constructor using localhost:11434 and llama3.1 model.
     */
    public OllamaClient() {
        this("http://localhost:11434", "llama3.1");
    }

    /**
     * Sends a prompt to the Ollama API and returns the generated response.
     *
     * @param prompt The full prompt to send to the model
     * @return The model's text response
     */
    public String generate(String prompt) throws IOException {
        JsonObject requestBody = new JsonObject();
        requestBody.addProperty("model", model);
        requestBody.addProperty("prompt", prompt);
        requestBody.addProperty("stream", false);

        // Optional: set generation parameters
        JsonObject options = new JsonObject();
        options.addProperty("temperature", 0.3);       // Lower temp for more factual summaries
        options.addProperty("num_predict", 2048);      // Max tokens in response
        options.addProperty("top_p", 0.9);
        requestBody.add("options", options);

        String jsonBody = gson.toJson(requestBody);

        Request request = new Request.Builder()
                .url(baseUrl + "/api/generate")
                .post(RequestBody.create(jsonBody, JSON_MEDIA))
                .build();

        logger.debug("Sending request to Ollama ({} model)...", model);

        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                String errorBody = response.body() != null ? response.body().string() : "No body";
                throw new IOException("Ollama API error (HTTP " + response.code() + "): " + errorBody);
            }

            String responseBody = response.body().string();
            JsonObject jsonResponse = gson.fromJson(responseBody, JsonObject.class);

            if (jsonResponse.has("response")) {
                return jsonResponse.get("response").getAsString();
            } else {
                throw new IOException("Unexpected response format: " + responseBody);
            }
        }
    }

    /**
     * Summarizes a FoxPro code block by building an appropriate prompt.
     *
     * @param codeBlock The FoxPro source code to analyze
     * @param context   Additional context (e.g., file name, module description)
     * @return The generated summary
     */
    public String summarizeFoxProCode(String codeBlock, String context) throws IOException {
        String prompt = buildSummarizationPrompt(codeBlock, context);
        return generate(prompt);
    }

    /**
     * Builds a structured prompt for FoxPro code summarization.
     */
    private String buildSummarizationPrompt(String code, String context) {
        return """
                You are a software analyst expert in Visual FoxPro (VFP) applications.
                Analyze the following FoxPro source code and provide a structured summary.

                Context: %s

                Please provide:
                1. **Use Case / Purpose**: What business function does this code serve?
                2. **Key Operations**: What are the main operations performed (CRUD, calculations, validations)?
                3. **Data Tables Used**: Which database tables (.dbf) are accessed and how?
                4. **Business Rules**: What validation rules or business logic is enforced?
                5. **User Interactions**: What inputs does it require and what outputs/messages does it produce?
                6. **Dependencies**: What other modules/procedures does it call?

                FoxPro Source Code:
                ```foxpro
                %s
                ```

                Provide a clear, concise summary suitable for documentation purposes.
                """.formatted(context, code);
    }

    /**
     * Tests connectivity to the Ollama server.
     *
     * @return true if server is reachable and model is available
     */
    public boolean testConnection() {
        try {
            Request request = new Request.Builder()
                    .url(baseUrl + "/api/tags")
                    .get()
                    .build();

            try (Response response = httpClient.newCall(request).execute()) {
                if (response.isSuccessful()) {
                    String body = response.body().string();
                    logger.info("Ollama server reachable. Available models: {}", body);
                    return body.contains(model);
                }
            }
        } catch (IOException e) {
            logger.error("Cannot connect to Ollama at {}: {}", baseUrl, e.getMessage());
        }
        return false;
    }

    public String getModel() {
        return model;
    }

    public String getBaseUrl() {
        return baseUrl;
    }
}
